sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableAsyncClient.java

```diff
@@ -1076,4 +1079,168 @@ private AccessPolicy toAccessPolicy(TableAccessPolicy tableAccessPolicy) {
+        TableServiceError error = null;
+        String errorMessage = null;
+        BatchChangeSet changes = null;
+        BatchOperation failedOperation = null;
+
+        if (requestBody.getContents().get(0) instanceof BatchChangeSet) {
+            changes = (BatchChangeSet) requestBody.getContents().get(0);
+        }
+
+        for (int i = 0; i < response.getValue().length; i++) {
+            BatchOperationResponse subResponse = response.getValue()[i];
+
+            // Attempt to attach a sub-request to each batch sub-response
+            if (changes != null && changes.getContents().get(i) != null) {
+                ModelHelper.updateBatchOperationResponse(subResponse,
+                    changes.getContents().get(i).getHttpRequest());
+            }
+
+            // If one sub-response was an error, we need to throw even though the service responded with 202
+            if (subResponse.getStatusCode() >= 400 && error == null && errorMessage == null) {
+                if (subResponse.getValue() instanceof TableServiceError) {
+                    error = (TableServiceError) subResponse.getValue();
+
+                    // Make a best effort to locate the failed operation and include it in the message
+                    if (changes != null && error.getOdataError() != null
+                        && error.getOdataError().getMessage() != null
+                        && error.getOdataError().getMessage().getValue() != null) {
+
+                        String message = error.getOdataError().getMessage().getValue();
+
+                        try {
+                            int failedIndex = Integer.parseInt(message.substring(0, message.indexOf(":")));
+                            failedOperation = changes.getContents().get(failedIndex).getOperation();
+                        } catch (NumberFormatException e) {
+                            // Unable to parse failed operation from batch error message - this just means
+                            // the service did not indicate which request was the one that failed. Since
+                            // this is optional, just swallow the exception.
+                        }
+                    }
+                } else if (subResponse.getValue() instanceof String) {
+                    errorMessage = "The service returned the following data for the failed operation: "
+                        + subResponse.getValue();
+                } else {
+                    errorMessage =
+                        "The service returned the following status code for the failed operation: "
+                            + subResponse.getStatusCode();
+                }
+            }
+        }
+
+        if (error != null || errorMessage != null) {
+            String message = "An operation within the batch failed, the transaction has been rolled back.";
+
+            if (failedOperation != null) {
+                message += " The failed operation was: " + failedOperation;
```

Comment:
Should we make the failedOperation value a property on the exception?

Reply:
I don't think so, we are already appending the contents of the failed operation in the exception message. It would look something like this:

```java
"CreateEntity{partitionKey='" + entity.getPartitionKey() + "', rowKey='" + entity.getRowKey() + "'}";
```

Reply:
Here is what we did in .NET after some discussion in the last arch board review:

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/tables/Azure.Data.Tables/src/TableTransactionFailedException.cs#L20

Reply:
Also, the failed operation type is the internal `BatchOperation`. We would need to either introduce a public type to represent these operations or just use their `String` representation. We would probably need to introduce a new type of exception, the one being used right now is very generic and used across the whole Tables SDK and I'm not a fan of adding a new property to it only for transactional operations.

Reply:
Ended up introducing a new type of exception for this: `TableFailedTransactionException`.